gitlab/resource_gitlab_project_protected_environment.go

```diff
@@ -0,0 +1,211 @@
+package gitlab
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+// https://docs.gitlab.com/ee/ci/environments/protected_environments.html
+func resourceGitlabProjectProtectedEnvironment() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGitlabProjectProtectedEnvironmentCreate,
+		Read:   resourceGitlabProjectProtectedEnvironmentRead,
+		Delete: resourceGitlabProjectProtectedEnvironmentDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"environment": {
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"deploy_access_levels": {
+				Type:     schema.TypeList,
+				ForceNew: true,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"access_level": {
+							Type:         schema.TypeString,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.StringIsNotEmpty,
+						},
+						"access_level_description": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"user_id": {
+							Type:         schema.TypeInt,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.StringIsNotEmpty,
+						},
+						"group_id": {
+							Type:         schema.TypeInt,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.StringIsNotEmpty,
+						},
+					},
```

Comment:
The schema validation could be improved here.

1. `user_id` and `group_id` should probably use integer validations. Since you don't use these attributes in your acceptance test, there's a good chance the `validation.StringIsNotEmpty` validation has a bug here.

2. The API docs say that exactly one of `access_level`, `user_id`, and `group_id` must be specified. You can check for that easily using `ExactlyOneOf`.

3. Usually for enum attributes like `access_level` we validate against the list of valid access levels (see https://github.com/gitlabhq/terraform-provider-gitlab/blob/9884585ddf50efa061798e452c1e3375b8714af7/gitlab/resource_gitlab_project_membership.go#L42)

Reply:
Regarding point 2, when trying to set `ExactlyOneOf`, there is no way to achieve that if multiple blocks are expected to be supported:

>```
>ExactlyOneOf: []string{"deploy_access_levels.0.acces_level", "deploy_access_levels.0.user_id", "deploy_access_levels.0.group_id"},
>ExactlyOneOf: []string{"deploy_access_levels.0.acces_level", "deploy_access_levels.0.user_id", "deploy_access_levels.0.group_id"},
>ExactlyOneOf: []string{"deploy_access_levels.0.acces_level", "deploy_access_levels.0.user_id", "deploy_access_levels.0.group_id"},
>```